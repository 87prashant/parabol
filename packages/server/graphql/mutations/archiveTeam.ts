import {GraphQLID, GraphQLNonNull} from 'graphql'
import getRethink from '../../database/rethinkDriver'
import ArchiveTeamPayload from '../types/ArchiveTeamPayload'
import {getUserId, isTeamLead} from '../../utils/authorization'
import publish from '../../utils/publish'
import sendSegmentEvent from '../../utils/sendSegmentEvent'
import shortid from 'shortid'
import standardError from '../../utils/standardError'
import {SubscriptionChannel} from 'parabol-client/types/constEnums'
import {NotificationEnum} from 'parabol-client/types/graphql'
import safeArchiveTeam from '../../safeMutations/safeArchiveTeam'

export default {
  type: ArchiveTeamPayload,
  args: {
    teamId: {
      type: new GraphQLNonNull(GraphQLID),
      description: 'The teamId to archive (or delete, if team is unused)'
    }
  },
  async resolve(_source, {teamId}, {authToken, dataLoader, socketId: mutatorId}) {
    const r = await getRethink()
    const now = new Date()
    const operationId = dataLoader.share()
    const subOptions = {operationId, mutatorId}

    // AUTH
    const viewerId = getUserId(authToken)
    if (!(await isTeamLead(viewerId, teamId))) {
      return standardError(new Error('Not team lead'), {userId: viewerId})
    }

    // RESOLUTION
    sendSegmentEvent('Archive Team', viewerId, {teamId}).catch()
    const {
      team,
      users,
      removedTeamNotifications,
      removedSuggestedActionIds
    } = await safeArchiveTeam(teamId)

    if (!team) {
      return standardError(new Error('Already archived team'), {userId: viewerId})
    }

    const notifications = users
      .map(({id}) => id)
      .filter((userId) => userId !== viewerId)
      .map((notifiedUserId) => ({
        id: shortid.generate(),
        startAt: now,
        type: NotificationEnum.TEAM_ARCHIVED,
        userIds: [notifiedUserId],
        teamId
      }))
    if (notifications.length) {
      await r
        .table('Notification')
        .insert(notifications)
        .run()
    }

    const data = {
      teamId,
      notificationIds: notifications.map(({id}) => id),
      removedTeamNotifications,
      removedSuggestedActionIds
    }
    publish(SubscriptionChannel.TEAM, teamId, 'ArchiveTeamPayload', data, subOptions)

    users.forEach((user) => {
      const {id, tms} = user
      publish(SubscriptionChannel.NOTIFICATION, id, 'AuthTokenPayload', {tms})
    })

    return data
  }
}
