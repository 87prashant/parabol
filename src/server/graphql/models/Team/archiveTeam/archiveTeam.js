import {GraphQLBoolean, GraphQLNonNull, GraphQLString} from 'graphql';
import getRethink from 'server/database/rethinkDriver';
import {auth0ManagementClient} from 'server/utils/auth0Helpers';
import {getUserId, requireSUOrLead, requireWebsocket} from 'server/utils/authorization';
import getPubSub from 'server/utils/getPubSub';
import sendSegmentEvent from 'server/utils/sendSegmentEvent';
import tmsSignToken from 'server/utils/tmsSignToken';
import shortid from 'shortid';
import {NEW_AUTH_TOKEN, NOTIFICATIONS_ADDED, TEAM_ARCHIVED} from 'universal/utils/constants';

export default {
  type: GraphQLBoolean,
  args: {
    teamId: {
      type: new GraphQLNonNull(GraphQLString),
      description: 'The teamId to archive (or delete, if team is unused)'
    }
  },
  async resolve(source, {teamId}, {authToken, socket}) {
    const r = getRethink();
    const now = new Date();

    // AUTH
    const userId = getUserId(authToken);
    const teamMemberId = `${userId}::${teamId}`;
    requireSUOrLead(authToken, teamMemberId);
    requireWebsocket(socket);

    // RESOLUTION
    sendSegmentEvent('Archive Team', userId, {teamId});
    const {teamResults: {notification}, userDocs} = await r.table('Team')
      .get(teamId)
      .pluck('name', 'orgId')
      .do((team) => ({
        projectCount: r.table('Project').getAll(teamId, {index: 'teamId'}).count(),
        team,
        userIds: r.table('TeamMember')
          .getAll(teamId, {index: 'teamId'})
          .filter({isNotRemoved: true})('userId')
          .coerceTo('array')
      }))
      .do((doc) => ({
        userDocs: r.table('User')
          .getAll(r.args(doc('userIds')), {index: 'id'})
          .update((user) => ({tms: user('tms').difference([teamId])}), {returnChanges: true})('changes')('new_val')
          .pluck('id', 'tms')
          .default([]),
        teamName: doc('team')('name'),
        teamResults: r.branch(
          r.and(doc('projectCount').eq(0), doc('userIds').count().eq(1)),
          {
            // Team has no projects nor addn'l TeamMembers, hard delete it:
            teamResult: r.table('Team').get(teamId).delete(),
            teamMemberResult: r.table('TeamMember').getAll(teamId, {index: 'teamId'}).delete()
          },
          {
            // Team has data or TeamMembers, archive team:
            notification: r.table('Notification').insert({
              id: shortid.generate(),
              orgId: doc('team')('orgId'),
              startAt: now,
              type: TEAM_ARCHIVED,
              userIds: doc('userIds'),
              teamName: doc('team')('name')
            }, {returnChanges: true})('changes')(0)('new_val').default(null),
            teamResult: r.table('Team').get(teamId).update({isArchived: true})
          }
        )
      }));

    if (notification) {
      const notificationsAdded = {notifications: [notification]};
      getPubSub().publish(`${NOTIFICATIONS_ADDED}.${teamId}`, {notificationsAdded});
    }
    userDocs.forEach((user) => {
      const {id, tms} = user;
      // update the tms on auth0 in async
      auth0ManagementClient.users.updateAppMetadata({id}, {tms});
      // update the server socket, if they're logged in
      getPubSub().publish(`${NEW_AUTH_TOKEN}.${id}`, {newAuthToken: tmsSignToken({sub: userId}, tms)});
    });

    return true;
  }
};
