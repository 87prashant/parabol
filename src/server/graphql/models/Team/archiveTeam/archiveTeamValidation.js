import legitify from 'universal/validation/legitify';
import {requiredId, teamName} from 'universal/validation/templates';

export default function addTeamValidation() {
  return legitify({
    id: requiredId,
    name: teamName,
    isArchived: (value) => value
  });
}
