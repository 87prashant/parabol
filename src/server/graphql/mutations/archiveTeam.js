import {GraphQLID, GraphQLNonNull} from 'graphql';
import getRethink from 'server/database/rethinkDriver';
import ArchiveTeamPayload from 'server/graphql/types/ArchiveTeamPayload';
import {auth0ManagementClient} from 'server/utils/auth0Helpers';
import {getUserId, requireTeamLead, requireWebsocket} from 'server/utils/authorization';
import getPubSub from 'server/utils/getPubSub';
import sendSegmentEvent from 'server/utils/sendSegmentEvent';
import tmsSignToken from 'server/utils/tmsSignToken';
import shortid from 'shortid';
import {NEW_AUTH_TOKEN, NOTIFICATIONS_ADDED, TEAM_ARCHIVED} from 'universal/utils/constants';
import toTeamMemberId from 'universal/utils/relay/toTeamMemberId';

const publishAuthTokensWithoutTeam = (userDocs) => {
  userDocs.forEach((user) => {
    const {id, tms} = user;
    // update the tms on auth0 in async
    auth0ManagementClient.users.updateAppMetadata({id}, {tms});
    // update the server socket, if they're logged in
    getPubSub().publish(`${NEW_AUTH_TOKEN}.${id}`, {newAuthToken: tmsSignToken({sub: id}, tms)});
  });
};

const publishTeamArchivedNotifications = (notifications, subOptions) => {
  notifications.forEach((notification) => {
    const userId = notification.userIds[0];
    const notificationsAdded = {notifications: [notification]};
    getPubSub().publish(`${NOTIFICATIONS_ADDED}.${userId}`, {notificationsAdded, ...subOptions});
  });
};

export default {
  type: ArchiveTeamPayload,
  args: {
    teamId: {
      type: new GraphQLNonNull(GraphQLID),
      description: 'The teamId to archive (or delete, if team is unused)'
    }
  },
  async resolve(source, {teamId}, {authToken, socket}) {
    const r = getRethink();
    const now = new Date();

    // AUTH
    const userId = getUserId(authToken);
    const teamMemberId = toTeamMemberId(teamId, userId);
    await requireTeamLead(teamMemberId);
    requireWebsocket(socket);

    // RESOLUTION
    sendSegmentEvent('Archive Team', userId, {teamId});
    const {teamResults: {notificationData, teamResult}, userDocs} = await r.table('Team')
      .get(teamId)
      .pluck('name', 'orgId')
      .do((team) => ({
        projectCount: r.table('Project').getAll(teamId, {index: 'teamId'}).count(),
        team,
        userIds: r.table('TeamMember')
          .getAll(teamId, {index: 'teamId'})
          .filter({isNotRemoved: true})('userId')
          .coerceTo('array')
      }))
      .do((doc) => ({
        userDocs: r.table('User')
          .getAll(r.args(doc('userIds')), {index: 'id'})
          .update((user) => ({tms: user('tms').difference([teamId])}), {returnChanges: true})('changes')('new_val')
          .pluck('id', 'tms')
          .default([]),
        teamName: doc('team')('name'),
        teamResults: r.branch(
          r.and(doc('projectCount').eq(0), doc('userIds').count().eq(1)),
          {
            // Team has no projects nor addn'l TeamMembers, hard delete it:
            teamResult: r.table('Team').get(teamId)
              .delete({returnChanges: true})('changes')(0)('old_val').default(null),
            teamMemberResult: r.table('TeamMember').getAll(teamId, {index: 'teamId'}).delete()
          },
          {
            notificationData: doc,
            teamResult: r.table('Team').get(teamId)
              .update({isArchived: true}, {returnChanges: true})('changes')(0)('new_val')
              .default(null)
          }
        )
      }));

    if (notificationData) {
      const {team: {name: teamName, orgId}, userIds} = notificationData;
      const notifications = userIds.map((notifiedUserId) => ({
        id: shortid.generate(),
        orgId,
        startAt: now,
        type: TEAM_ARCHIVED,
        userIds: [notifiedUserId],
        teamName
      }));
      await r.table('Notification').insert(notifications);
      publishTeamArchivedNotifications(notifications);
    }
    publishAuthTokensWithoutTeam(userDocs);

    if (!teamResult) {
      throw new Error('Team was already archived');
    }
    return {team: teamResult};
  }
};
