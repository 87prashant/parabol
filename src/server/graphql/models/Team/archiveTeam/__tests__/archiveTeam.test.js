import DynamicSerializer from 'dynamic-serializer';
import MockDate from 'mockdate';
import getRethink from 'server/database/rethinkDriver';
import {__now} from 'server/__tests__/setup/mockTimes';
import fetchAndSerialize from 'server/__tests__/utils/fetchAndSerialize';
import MockDB from 'server/__tests__/setup/MockDB';
import archiveTeam from 'server/graphql/models/Team/archiveTeam/archiveTeam';
import mockAuthToken from 'server/__tests__/setup/mockAuthToken';
import socket from 'server/__mocks__/socket';

MockDate.set(__now);
console.error = jest.fn();

describe('ArchiveTeam', () => {
  test('archives a team', async () => {
    // SETUP
    const r = getRethink();
    const dynamicSerializer = new DynamicSerializer();
    const mockDB = new MockDB();
    const {user, team: [updatedTeam], teamMember} = await mockDB.init();
    updatedTeam.isArchived = true;
    updatedTeam.name = updatedTeam.teamName;
    const teamLeadId = teamMember.find((tm) => tm.teamId === updatedTeam.id && tm.isLead).userId;
    const teamLead = user.find((usr) => usr.id === teamLeadId);
    const authToken = mockAuthToken(teamLead);

    // TEST
    await archiveTeam.resolve(undefined, {updatedTeam}, {authToken, socket});

    // VERIFY
    const db = await fetchAndSerialize({
      team: r.table('Team').get(updatedTeam.id),
      notification: r.table('Notification').getAll(updatedTeam.orgId, {index: 'orgId'}).orderBy('startAt')
    }, dynamicSerializer);
    expect(db).toMatchSnapshot();
  });
});
