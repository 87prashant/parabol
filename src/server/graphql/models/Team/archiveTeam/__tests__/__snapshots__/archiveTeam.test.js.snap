// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArchiveTeam archives a team 1`] = `
Object {
  "notification": Array [
    Object {
      "id": 2,
      "orgId": 1,
      "startAt": 2017-01-01T08:00:00.000Z,
      "type": "TEAM_ARCHIVED",
      "userIds": Array [
        3,
        4,
        5,
        6,
        7,
        8,
        9,
        10,
      ],
      "varList": Array [
        "Team America",
      ],
    },
  ],
  "team": Object {
    "activeFacilitator": null,
    "facilitatorPhase": "lobby",
    "facilitatorPhaseItem": null,
    "id": 0,
    "isArchived": true,
    "isPaid": true,
    "meetingId": null,
    "meetingPhase": "lobby",
    "meetingPhaseItem": null,
    "orgId": 1,
    "teamName": "Team America",
  },
}
`;
